### PR TITLE
[tools/onert_train] Remove default value from onert_train/args

### PR DESCRIPTION
**Pull request by zetwhite**
This PR removes default value from onert_train's Args class. This is to use model file parameter as default value, if nothing is given from command line input.

ONE-DCO-1.0-Signed-off-by: SeungHui Youn <sseung.youn@samsung.com>

related with: https://github.com/Samsung/ONE/issues/12520

### CODE REVIEW

File: tests/tools/onert_train/src/onert_train.cc [hunk 1]
```diff
@@ -138,11 +138,16 @@ int main(const int argc, char **argv)
...
+    if (tri.batch_size == 0)
+    {
+      // If args doens't have a batch size, set batch size 1
+      tri.batch_size = args.getBatchSize().value_or(1);
+    }
+    tri.learning_rate = args.getLearningRate().value_or(tri.learning_rate);
+    tri.loss_info.loss = args.getLossType().value_or(tri.loss_info.loss);
+    tri.loss_info.reduction_type =
+      args.getLossReductionType().value_or(tri.loss_info.reduction_type);
+    tri.opt = args.getOptimizerType().value_or(tri.opt);
```
Comment (zetwhite) [inline tests/tools/onert_train/src/onert_train.cc:150]: `value_or` :  
  * If optional has value, the value inside the optional class is returned.
  * If optional doesn't have a value, the value written in `value_or()` is returned.

ref : https://en.cppreference.com/w/cpp/utility/optional/value_or

File: tests/tools/onert_train/src/onert_train.cc [hunk 2]
```diff
@@ -138,11 +138,16 @@ int main(const int argc, char **argv)
...
     // overwrite training information using the arguments
-    tri.batch_size = args.getBatchSize();
-    tri.learning_rate = args.getLearningRate();
-    tri.loss_info.loss = args.getLossType();
-    tri.loss_info.reduction_type = args.getLossReductionType();
-    tri.opt = args.getOptimizerType();
+    if (tri.batch_size == 0)
+    {
+      // If args doens't have a batch size, set batch size 1
+      tri.batch_size = args.getBatchSize().value_or(1);
```
Comment (zetwhite) [inline tests/tools/onert_train/src/onert_train.cc:144]: if both the model file and cmd-line options doesn't give a batch size, 
Still make it run somehow, by setting batch_size as 1.

**Comment by YongseopKim** [comment]
Q. What does `model file parameter` mean?

**Comment by zetwhite** [comment]
> Q. What does `model file parameter` mean?

@YongseopKim 

I'd like to express that if nothing is given a value in the circle file is chosen.
Training Information(parameter) including batch size, learning rate, optimizer, loss are saved in circle file's metadata. 

About circle file's metadata : [:link:](https://github.sec.samsung.net/one-project/circle-x/blob/main/spec/circle%2Bspec.rst#41-circle_training)

**Comment by YongseopKim** [comment]
@nnfw-bot test onert-cross-release

**Comment by zetwhite** [comment]
### For reviewers

Here is a overall policy in onert_train. 

* `onert_train` expect that model file has training information. 
* If a user want to run onert_train by setting training information manually, user can update training information by cmd line option.


| Case | Model file info is given | cmd line info is given | onert_train Selected | 
| ----     | ------                             |  -----------                    | --------      |
| 1        | X                                  | X                               |  train info is undefined, throw error | 
| 2        | O                                  | X                               |  model file info selected | 
| 3        | O                                 | O                               | cmd line info is selected |

**Comment by YongseopKim** [comment]
The word `Model file info` is ambiguous. It can mean such as file size, file path and etc. `Training info` is better or such as `Model hyper-parameter` is a good option because the word `model hyper param` has been used in so many places.

File: tests/tools/onert_train/src/onert_train.cc [hunk 3]
```diff
@@ -138,11 +138,16 @@ int main(const int argc, char **argv)
     NNPR_ENSURE_STATUS(nnfw_train_get_traininfo(session, &tri));
 
     // overwrite training information using the arguments
-    tri.batch_size = args.getBatchSize();
-    tri.learning_rate = args.getLearningRate();
-    tri.loss_info.loss = args.getLossType();
-    tri.loss_info.reduction_type = args.getLossReductionType();
-    tri.opt = args.getOptimizerType();
+    if (tri.batch_size == 0 && !args.getBatchSize().has_value())
```
Comment (YongseopKim) [inline tests/tools/onert_train/src/onert_train.cc:141]: AFAIU, now `tri` has train info including `batch_size`, `learning_rate` and ...

File: tests/tools/onert_train/src/onert_train.cc [hunk 4]
```diff
@@ -138,11 +138,16 @@ int main(const int argc, char **argv)
...
+      tri.batch_size = 1;
+    else
+      tri.batch_size = args.getBatchSize().value_or(tri.batch_size);
+
+    tri.learning_rate = args.getLearningRate().value_or(tri.learning_rate);
+    tri.loss_info.loss = args.getLossType().value_or(tri.loss_info.loss);
+    tri.loss_info.reduction_type =
+      args.getLossReductionType().value_or(tri.loss_info.reduction_type);
+    tri.opt = args.getOptimizerType().value_or(tri.opt);
```
Comment (YongseopKim) [inline tests/tools/onert_train/src/onert_train.cc:150]: Okay, after this line `tri`'s values are filled sth(it could be valid or invalid.) If these values are invalid, nnfw_api throws error.

File: tests/tools/onert_train/src/args.cc [hunk 5]
```diff
@@ -113,9 +113,9 @@ T checkValidation(const std::string &arg_name, const std::vector<T> &valid_args,
 
 // generate a help message based on the valid_args and default_arg
 template <typename T>
-std::string genHelpMsg(const std::string &arg_name, const std::vector<T> &valid_args, T default_arg)
+std::string genHelpMsg(const std::string &arg_name, const std::vector<T> &valid_args)
 {
-  std::string msg = arg_name + " (default: " + onert_train::to_string(default_arg) + ")\n";
+  std::string msg = arg_name + " (default: use model file parameter)\n";
```
Comment (zetwhite) [inline tests/tools/onert_train/src/args.cc:118]: Some other candidate words to use here.

```suggestion
  std::string msg = arg_name + " (If nothing is given, model hyper param is selected)\n";
```

File: tests/tools/onert_train/src/args.cc [hunk 6]
```diff
@@ -251,22 +251,19 @@ void Args::Initialize(void)
     )
     ("mem_poll,m", po::value<bool>()->default_value(false)->notifier([&](const auto &v) { _mem_poll = v; }), "Check memory polling (default: false)")
     ("epoch", po::value<int>()->default_value(5)->notifier([&](const auto &v) { _epoch = v; }), "Epoch number (default: 5)")
-    ("batch_size", po::value<int>()->default_value(32)->notifier([&](const auto &v) { _batch_size = v; }), "Batch size (default: 32)")
-    ("learning_rate", po::value<float>()->default_value(0.001)->notifier([&](const auto &v) { _learning_rate = v; }), "Learning rate (default: 0.001)")
+    ("batch_size", po::value<int>()->notifier([&](const auto &v) { _batch_size = v; }), "Batch size (default: use model file parameter)")
```
Comment (YongseopKim) [inline tests/tools/onert_train/src/args.cc:254]: I talked about the word `default` with @zetwhite . However, I can't imagine the alternative of the `default`. Therefore, I will not disagree that the word `default` keeps used.

File: tests/tools/onert_train/src/onert_train.cc (hunk 3 above)
Comment (YongseopKim) [reply tests/tools/onert_train/src/onert_train.cc:141]: Why is `batch_size` being handled differently?

File: tests/tools/onert_train/src/args.cc (hunk 5 above)
Comment (zetwhite) [reply tests/tools/onert_train/src/args.cc:118]: @YongseopKim How about this word? Is it a bit better or not..?

Comment (YongseopKim) [reply tests/tools/onert_train/src/args.cc:118]: Great!

File: tests/tools/onert_train/src/onert_train.cc (hunk 3 above)
Comment (zetwhite) [reply tests/tools/onert_train/src/onert_train.cc:141]: @jyoungyun gives me a idea about batch_size default value in the draft : https://github.com/Samsung/ONE/pull/12445/files#r1461482143  

> (suggestion)
> How about setting a default value for batch_size? Since the batch size is not related to the model file, onert_train can provide a default batch size value to the user.

> TensorFlow sets 32 as the default batch size, and PyTorch sets 1 as the default batch size. In ONERT, if the data length is not divisible by the batch size, it does not train the rest of the dataset. Therefore, I suggest setting 1 as the batch size for onert_train. 


Hmm.. but rethinking this part, I'd better talk about this with @jyoungyun once again.

Comment (zetwhite) [reply tests/tools/onert_train/src/onert_train.cc:141]: Rethinking about this part, I changed my mind a bit. 

To give a default value to the batch_size, Maybe we'd better update IR::TrainingInfo has a default value. 
https://github.com/Samsung/ONE/blob/356901d7a13c270252919834b581096e2cf35663/runtime/onert/core/include/ir/train/TrainingInfo.h#L35 

To give a default value on batch_size, there are places to inject the default value. 
 * make onert_train/Args has a default value
     But if we set a default value on Args class. It is hard to distinguish default value and user given value. 

 * Second, make onert_train.cc set a default value if both model file and cmd opt aren't given 
     But it just changes one tool's policy, not the training core. 

  * Third, update ir::TrainingInfo constructor to set a default value.
     This looks best to me.

Comment (zetwhite) [reply tests/tools/onert_train/src/onert_train.cc:141]: @jyoungyun I'd be grateful if you left any opinion about this. 
Since it is tricky to discuss online, I'm also welcome to have some talk on offline :smile:

File: tests/tools/onert_train/src/args.cc [hunk 7]
```diff
@@ -251,22 +252,23 @@ void Args::Initialize(void)
...
       ->notifier([&](const auto &v){_loss_reduction_type = checkValidation("loss_reduction_type", valid_loss_rdt, v);}),
-      genHelpMsg("loss_reduction_tye", valid_loss_rdt, default_loss_rdt).c_str()
+      genHelpMsg("Loss Reduction type", valid_loss_rdt).c_str()
     )
     ("optimizer", po::value<int>()
-      ->default_value(default_optim)
       ->notifier([&](const auto& v){_optimizer_type = checkValidation("optimizer", valid_optim, v);}),
-      genHelpMsg("optimizer", valid_optim, default_optim).c_str()
+      genHelpMsg("Optimizer type", valid_optim).c_str()
     )
```
Comment (zetwhite) [inline tests/tools/onert_train/src/args.cc:272]: Help message example : 

```bash
➜ ~/Workspace/ONE_second/Product/build/bin/onert_train --help
onert_train

Usage: /home/sseung/Workspace/ONE_second/Product/build/bin/onert_train [model path] [<options>]


General options:
  -h [ --help ]                    Print available options
  --version                        Print version and exit immediately
  --nnpackage arg                  NN Package file(directory) name
  --modelfile arg                  NN Model filename
  --path arg                       NN Package or NN Modelfile path
  --export_path arg                Path to export circle
  --load_input:raw arg             NN Model Raw Input data file
                                   The datafile must have data for each input number.
                                   If there are 3 inputs, the data of input0 must exist as much as 
                                   data_length, and the data for input1 and input2 must be held 
                                   sequentially as data_length.
  --load_expected:raw arg          NN Model Raw Expected data file
                                   (Same data policy with load_input:raw)
  -m [ --mem_poll ] arg (=0)       Check memory polling (default: false)
  --epoch arg (=5)                 Epoch number (default: 5)
  --batch_size arg                 Batch size
                                   If not given, model's hyper parameter is used
  --learning_rate arg              Learning rate
                                   If not given, model's hyper parameter is used
  --loss arg                       Loss type
                                   If not given, model's hyper parameter is used
                                   1: mean squared error
                                   2: categorical crossentropy
  --loss_reduction_type arg        Loss Reduction type
                                   If not given, model's hyper parameter is used
                                   1: sum over batch size
                                   2: sum
  --optimizer arg                  Optimizer type
                                   If not given, model's hyper parameter is used
                                   1: sgd
                                   2: adam
  --metric arg (=-1)               Metric type
                                   Simply calculates the metric value using the variables (default:
                                   none)
                                   0: CATEGORICAL_ACCURACY
  --validation_split arg (=0)      Float between 0 and 1(0 < float < 1). Fraction of the training 
                                   data to be used as validation data.
  -v [ --verbose_level ] arg (=0)  Verbose level
                                   0: prints the only result. Messages btw run don't print
                                   1: prints result and message btw run
                                   2: prints all of messages to print
  --output_sizes arg               The output buffer size in JSON 1D array
                                   If not given, the model's output sizes are used
                                   e.g. '[0, 40, 2, 80]' to set 0th tensor to 40 and 2nd tensor to 
                                   80.
```

File: tests/tools/onert_train/src/onert_train.cc (hunk 3 above)
Comment (jyoungyun) [reply tests/tools/onert_train/src/onert_train.cc:141]: > (suggestion)
> How about setting a default value for batch_size? Since the batch size is not related to the model file, onert_train can provide a default batch size value to the user.
> 
> TensorFlow sets 32 as the default batch size, and PyTorch sets 1 as the default batch size. In ONERT, if the data length is not divisible by the batch size, it does not train the rest of the dataset. Therefore, I suggest setting 1 as the batch size for onert_train.

A few weeks ago, I suggested to use default value for the batch_size. This is because it is easier to use the `onert_train` tool for users. However, if the model file is circle file rather than circle+ file, the user must provide many parameters. I think it is better that the batch_size is handled like other paramters. Let's remove the default value of batch_size!

**Comment by zetwhite** [comment]
@YongseopKim @jyoungyun PTAL, Thank you in advance. :smile:

**Review by YongseopKim — APPROVED** [review]
LGTM 👍
